File: Leet Code/Linked List Cycle II/Linked_List_Cycle_II.cpp
// Approach 1 :

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectLoop(ListNode *head){
        ListNode *slow = head, *fast = head;
        
        while(fast != NULL && fast->next != NULL){
            slow = slow->next;
            fast = fast->next->next;
            
            if(slow == fast){
                return slow;
            }
        }
        return NULL;
    }
    ListNode *detectCycle(ListNode *head) {
        if(head == NULL || detectLoop(head) == NULL){
            return NULL;
        }
        
        ListNode *start = head;
        ListNode *last = detectLoop(head);
        
        while(start != last){
            start = start->next;
            last = last->next;
        }
        return start;
    }
};



// Approach 2 :

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {

        if(head == NULL){
            return NULL;
        }

        ListNode *low = head, *high = head;
        bool cycle = false;
        
        while(high != NULL && high->next != NULL){

            low = low->next;
            high = high->next->next;

            if(low == high){
                cycle = true;
                break;
            }
        }
        
        if(!cycle){
            return NULL;
        }
        else{
            high = head;
            while(low != high){
                low = low->next;
                high = high->next;
            }
            return low;
        }
    }
};

File: Leet Code/Linked List Cycle II/Linked_List_Cycle_II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {

        if(head == null){
            return null;
        }

        ListNode slow = head, fast = head;
        boolean flag = false;

        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;

            if(slow == fast){
                flag = true;
                break;
            }
        }

        if(!flag){
            return null;
        }
        else{
            fast = head;
            while(slow != fast){
                slow = slow.next;
                fast = fast.next;
            }
            return slow;
        }

    }
}
